🎉 New Source: Chartmogul

TSkrebe opened the pull request:
## What
This PR adds a new source: [chartmogul api](https://dev.chartmogul.com/)


## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [x] Connector's `README.md`
    - [x] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/SUMMARY.md`
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>


<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/airbytehq/airbyte/9381)
<!-- Reviewable:end -->

TSkrebe commented:
./acceptance-test-docker.sh
```
 test_core.py ✓✓✓✓✓✓✓✓✓✓✓✓✓                                                                                                                                                                                        93% █████████▍
 test_full_refresh.py ✓                                                                                                                                                                                           100% ██████████
==================================================================================================== short test summary info ====================================================================================================
SKIPPED [1] source_acceptance_test/plugin.py:56: Skipping TestIncremental.test_two_sequential_reads because not found in the config

Results (73.78s):
      14 passed
```

TSkrebe commented:
Not sure how to proceed with `airbyte-integrations/builds.md`

marcosmarxm reviewed (approved):
The connector looks well implemented, I created one issue to get credentials to apply CI tests; If possible to share a testing credentials this helps to merge sooner.

marcosmarxm commented:
> Not sure how to proceed with `airbyte-integrations/builds.md`

The link is built after the connector is merged.

karinakuz commented:
Account with the data is created, credentials are added to the LastPass with the name "Chartmogul", api is tested for the strams:
- Get Data Sources https://api.chartmogul.com/v1/data_sources
- Get Customers https://api.chartmogul.com/v1/customers
- Get Plans https://api.chartmogul.com/v1/plans
- Get Invoices https://api.chartmogul.com/v1/import/customers/cus_e6e6b1e6-72d4-11ec-ac99-47c8cf90d52c/invoices

TSkrebe commented:
i'm in no rush, just curious on ETA when we could see this published?

alafanechere commented:
Hi @TSkrebe, we created a sandbox account for Chartmogul. I'll now open a side branch to run the acceptance tests on this account. If they pass I'll go for a review, and if I've no additional change request we can publish it today the beginning of next week.

alafanechere commented:
Hi @TSkrebe I made some minor changes to make the acceptance tests pass:
* Fix some unused import to make Flake check pass
* Extensive use of mocker fixture in unit tests to replace mock.MagicMock with mocker.MagicMock
* **Change customer id field to integer as it's the type the API returned**
* **Make primary key fields not nullable in streams schema. Your schemas have a lot of nullable field, please double-check they are all really nullable according to Chartmogul API reference.**
* **Declare `activities` as an empty stream in `basic_read` acceptance test as our sandbox account does not have any activity. Do you know how we could generate an activity? I'm not familiar with Chartmogul.**

I highlighted the changes on which I'd like a reply from your side.

alafanechere commented:
> Not sure how to proceed with airbyte-integrations/builds.md

You have to add a line for your connector at the right alphabetical place and copy what's been made for other connectors.
I did it in https://github.com/airbytehq/airbyte/pull/9381/commits/bd82338c454e3e5cc14c13e1dd3ce23869812599

alafanechere commented on airbyte-integrations/connectors/source-chartmogul/source_chartmogul/source.py:
```diff
@@ -0,0 +1,94 @@
+
+
+class Customers(ChartmogulStream):
+    primary_key = "uuid"
```
Customer schema has an `id` and an `uuid` field, what makes `uuid` more legit than `id` for a primary key?

TSkrebe replied:
good question. Both seem to be unique according to documentation, but i picked uuid since, well, it's uuid and should always be unique.. Not the best argument haha 

Is there any performance impact on string vs int primary key?

alafanechere replied:
int is probably better for performance in for a PK (easier to sort and lighter)

alafanechere commented on airbyte-integrations/connectors/source-chartmogul/source_chartmogul/source.py:
```diff
@@ -0,0 +1,94 @@
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        json_response = response.json()
+        if not json_response.get("has_more", False):
+            return None
+
+        return {"page": json_response.get("current_page") + 1}
```
nit, for readability:
```suggestion
        if json_response.get("has_more", False):
            return {"page": json_response.get("current_page") + 1}

        return None
```

alafanechere commented on airbyte-integrations/connectors/source-chartmogul/source_chartmogul/source.py:
```diff
@@ -0,0 +1,94 @@
+        return "v1/customers"
+
+
+class Activities(ChartmogulStream):
```
I think this stream might be a good candidate for an incremental sync mode as you can manipulate the `start_date` and `end_date` parameters. Is there any limitation that prevents you from implementing incremental logic? 

You can do this in another PR if you feel it's too much additional work.

TSkrebe replied:
It's tricky to make a solid version. Decided not to include in initial PR. Copy-pasting from bootstrap.md:
> One would need to keep track of both UUID and created_at and read stream until `datetime.now()`. Dynamic end date would be necessary since activities can also have a future date. Since data can also be changed retrospectively, a `lookback window` would also be necessary to catch all the changes.

alafanechere commented on airbyte-integrations/connectors/source-chartmogul/source_chartmogul/source.py:
```diff
@@ -0,0 +1,94 @@
+        params = {}
+
+        if next_page_token:
+            params["start-after"] = next_page_token["start-after"]
```
```suggestion
            params.update(next_page_token)
```

alafanechere commented on airbyte-integrations/connectors/source-chartmogul/unit_tests/test_source.py:
```diff
@@ -0,0 +1,28 @@
+
+    # failure
+    requests_mock.get("https://api.chartmogul.com/v1/ping", status_code=500)
+    assert source.check_connection(logger_mock, config_mock) == (False, ANY)
```
I think the error is of type `HTTPError` if you used `raise_for_status`, you could test the error is of this type.

alafanechere commented on airbyte-integrations/connectors/source-chartmogul/unit_tests/test_streams.py:
```diff
@@ -0,0 +1,116 @@
+    mocker.patch.object(Activities, "__abstractmethods__", set())
+
+
+# Customer stream tests
```
You can group testing of specific classes in test classes:
eg. 
```
class TestCustomers:
    def test_requests_params(self):
        stream = Customers()
```
It would avoid the awkward suffix naming for tests such as `test_request_params_activities`

alafanechere commented on airbyte-integrations/connectors/source-chartmogul/unit_tests/test_streams.py:
```diff
@@ -0,0 +1,116 @@
+    assert stream.next_page_token(**inputs) == {"start-after": "unique-uuid"}
+
+
+# Default tests
+
+
+@pytest.mark.parametrize(
+    ("http_status", "should_retry"),
+    [
+        (HTTPStatus.OK, False),
+        (HTTPStatus.BAD_REQUEST, False),
+        (HTTPStatus.TOO_MANY_REQUESTS, True),
+        (HTTPStatus.INTERNAL_SERVER_ERROR, True),
+    ],
+)
+def test_should_retry(http_status, should_retry, mocker):
+    response_mock = mocker.MagicMock()
+    response_mock.status_code = http_status
+    stream = Customers()
+    assert stream.should_retry(response_mock) == should_retry
+
+
+def test_backoff_time(mocker):
+    response_mock = mocker.MagicMock()
+    stream = Customers()
+    expected_backoff_time = None
+    assert stream.backoff_time(response_mock) == expected_backoff_time
```
You can remove these tests if you did not override these parent class methods.

alafanechere reviewed (changes requested):
Thank you @TSkrebe, the overall code looks good to me. I still have some minor suggestions. Let me know what you think.

TSkrebe commented:
@alafanechere 

>Change customer id field to integer as it's the type the API returned

makes sense, don't why it was a string. thanks

>Declare activities as an empty stream in basic_read acceptance test. Our sandbox account does not have any activity. Do you know how we could generate an activity? I'm not familiar with Chartmogul.

in my case, there is integration with stripe https://app.chartmogul.com/#/data-platform/data-sources . Whenever there is a payment, a new activity is created. If there is any activities in your stripe sandbox it should theoretically work.

>Make primary key fields not nullable in streams schema. Your schemas have a lot of nullable fields, please double-check they are all really nullable according to Chartmogul API reference.

Since they don't provide open api, it's difficult to know which ones can and can't be null.  


thanks for the feedback on python style. Difficult to switch from golang midset... will address after the weekend.

TSkrebe commented:
@alafanechere addressed your comments. let me know if i missed anything. you'll probably need to run gradlew command again as i wasn't able to get it up-and-running on my M1

alafanechere reviewed (approved):
Thanks for the changes you made. I confirm that acceptance test are passing locally for me after I made some unused import removal that blocked the flake check.
